wordpress-mobile/WordPress-FluxC-Android
# [Dependency Analysis] Add Android Gradle Plugin

Project Thread: paaHJt-6yU-p2
Required By: [BuildkiteCI#484](https://github.com/Automattic/buildkite-ci/pull/484)
Depends On:
 - [x] [ADC#31](https://github.com/Automattic/android-dependency-catalog/pull/31)
 - [x] Update ADC to `1.19.0`

This PR adds the dependency analysis Android Gradle plugin to this project, for dependency analysis purposes.

### Description


For now, only the main `buildHealth` task is going to be utilized and produce data once every week on CI (see this [commit](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/3063/commits/c412ac31eb7e8bcd06d245d35fd7e9e401acac19) and [this](https://github.com/Automattic/buildkite-ci/pull/484) PR). Amongst other, this data will include the following:

- Number of modules (`projectCount`)
- Unused dependencies which should be removed (`unusedCount`)
- Transitive dependencies which should be declared directly (`undeclaredCount`)
- Existing dependencies which should be modified to be as indicated (`misDeclaredCount`)
- Dependencies which could be compile-only (`compileOnlyCount`)
- Dependencies which should be removed or changed to runtime-only (`runtimeOnlyCount`)

Afterward, this data will get collected from CI and uploaded to our Apps Metrics infrastructure, for visualization and alerting purposes.

### Testing Steps

1. Local: Run the `./gradlew buildHealth` task and verify that under the root level `build/reports/dependency-analysis` folder you get the below 2 reports both, in JSON and text format:
    - `build-health-report.json`
    - `build-health-report.txt`
2. CI: Using the `New Build` 🟢 CI button for [FluxC](https://buildkite.com/automattic/wordpress-fluxc-android), test this standalone dependency analysis job (see form below). Then:
    - When CI starts, make sure that adding `/meta-data` to that CI build's URL ([example](https://buildkite.com/automattic/wordpress-fluxc-android/builds/3744/meta-data)) will give you one extra meta-data, the `pipeline_file` one, with a value of `schedules/dependency-analysis.yml` (see screenshot below):
    - When CI completes, make sure that the below 2 CI artifacts are available both, in JSON and text format:
        - `build/reports/dependency-analysis/build-health-report.json`
        - `build/reports/dependency-analysis/build-health-report.txt`
3. REST: Using the Buildkite's REST API you could query for this type of builds, using the [meta-data](https://buildkite.com/docs/apis/rest-api/builds#list-all-builds) query parameter, and see that the result is the expect one. Use this query: `https://api.buildkite.com/v2/organizations/automattic/pipelines/wordpress-fluxc-android/builds?meta_data[pipeline_file]=schedules/dependency-analysis.yml`

![image](https://github.com/user-attachments/assets/c5f69bbc-ff77-40f4-b609-11bc90a93e8e)

![image](https://github.com/user-attachments/assets/0be39263-95e6-40db-9df0-bfb64d203950)
